query/aggregate.go

```diff
@@ -0,0 +1,192 @@
+type aggregate int
+
+const (
+	sumAggregate aggregate = iota + 1
+	meanAggregate
+	minAggregate
+	maxAggregate
+)
+
+var aggregatorMap = map[aggregate]func([]float64) float64{
+	sumAggregate:
+	// The aggregatefor sum finds the sum of the given array.
+	func(array []float64) float64 {
+		sum := 0.0
+		for _, v := range array {
+			sum += v
+		}
+		return sum
+	},
+
+	meanAggregate:
+	// The mean aggregator returns the mean of the given array
+	func(array []float64) float64 {
+		if len(array) == 0 {
+			// The mean of an empty list is not well-defined
+			return math.NaN()
+		}
+		sum := 0.0
+		for _, v := range array {
+			sum += v
+		}
+		return sum / float64(len(array))
+	},
+	minAggregate:
+	// The minimum aggregator returns the minimum
+	func(array []float64) float64 {
+		if len(array) == 0 {
+			// The minimum of an empty list is not well-defined
+			return math.NaN()
+		}
+		min := array[0]
+		for _, v := range array {
+			min = math.Min(min, v)
+		}
+		return min
+	},
+	maxAggregate:
+	// The maximum aggregator returns the maximum
+	func(array []float64) float64 {
+		if len(array) == 0 {
+			// The maximum of an empty list is not well-defined
+			return math.NaN()
+		}
+		max := array[0]
+		for _, v := range array {
+			max = math.Max(max, v)
+		}
+		return max
+	},
+}
+
+// applyAggregation takes an aggregation function ( [float64] => float64 ) and applies it to a given list of Timeseries
+// the list must be non-empty, or an error is returned
+func applyAggregation(group group, aggregator func([]float64) float64) api.Timeseries {
+	list := group.List
+	tagSet := group.TagSet
+
+	if len(list) == 0 {
+		// This case cannot actually occur, provided the rest of the code has been implemented correctly.
+		// However, if it *does* occur, then we need to exit early to prevent a panic when we access list[0].
+		return api.Timeseries{
+			Values: []float64{},
+			TagSet: tagSet,
+		}
+	}
+
+	series := api.Timeseries{
+		Values: make([]float64, len(list[0].Values)), // The first Series in the given list is used to determine this length
+		TagSet: tagSet,                               // The tagset is supplied by an argument (it will be the values grouped on)
+	}
+
+	// Make a slice of time to reuse.
+	// Each entry corresponds to a particular Series, all having the same index within their corresponding Series.
+	timeSlice := make([]float64, len(list))
+
+	for i := range series.Values {
+		// We need to determine each value in turn.
+		for j := range timeSlice {
+			timeSlice[j] = list[j].Values[i]
+		}
+		// Find the aggregated value:
+		series.Values[i] = aggregator(timeSlice)
+	}
+
+	return series
+}
+
+// This function is the culmination of all others.
+// `aggregateBy` takes a series list, an aggregator, and a set of tags.
+// It produces a SeriesList which is the result of grouping by the tags and then aggregating each group
+// into a single Series.
+func aggregateBy(list api.SeriesList, aggregate aggregate, tags []string) api.SeriesList {
```

Comment:
i think it's better if we pass in `func([]float64)float64` directly from the outside. they can do the enum-to-function mapping.

(one better thing about that approach is that we can make this file more generic, and additional functions can be added in tests, etc without modifying this file).